### PR TITLE
OCPQE-18693: automate azure disk encryption set test cases

### PR DESCRIPTION
**Pull request by jinyunma**
OCP-60977 - [ipi-on-azure] install cluster with different custom managed key setting for master and worker

1.  update des related steps (provision, grant-permission, post-check, ipi-conf-provisioned-des) to support different des set under controlplane/compute/defalutMachinePlatform
2.  update job `azure-ipi-des-minimal-permission-amd-f28-destructive` to support different des for master and worker
3.  update step `azure-provision-disk-encryption-set-grant-permission` to consider `ENABLE_MIN_PERMISSION_FOR_DES`, and grant the correct permissions.

### CODE REVIEW

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.15-azure-ipi-des-master-worker-f28

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-des-minimal-permission-arm-f14

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.15-azure-ipi-des-diffrent-on-worker-f28

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.15-azure-ipi-des-diffrent-on-worker-f28

File: ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-commands.sh [hunk 1]
```diff
@@ -75,12 +75,27 @@ keyvault="${NAMESPACE}-${UNIQUE_HASH}-kv"
 keyvault_key="${NAMESPACE}-${UNIQUE_HASH}-kvkey"
 des="${NAMESPACE}-${UNIQUE_HASH}-des"
 create_disk_encryption_set ${RESOURCE_GROUP} ${keyvault} ${keyvault_key} ${des}
+des_list="${des}"
 
-echo "Granting service principal reader permissions to the DiskEncryptionSet: ${des}"
-des_id=$(az disk-encryption-set show -n ${des} -g ${RESOURCE_GROUP} --query "[id]" -o tsv)
-cluster_sp_id=$(cat "${AZURE_AUTH_LOCATION}" | jq -r ".clientId")
-run_command "az role assignment create --assignee ${cluster_sp_id} --role Owner --scope ${des_id} -o jsonc"
+if [[ "${SET_DIFFERENT_DES_FOR_COMPUTE}" == "true" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-commands.sh:80]: Some days ago ibmcloud did some similar functions, refer to https://github.com/openshift/release/pull/47656, that means, once any of `ENABLE_DES_DEFAULT_MACHINE`, `ENABLE_DES_COMPUTE`, and `ENABLE_DES_CONTROL_PLANE` is enabled, always create a dedicated DES, than we do not need to introduce `SET_DIFFERENT_DES_FOR_COMPUTE` any more.

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-commands.sh]: updated.
seems that OCP-51422 is duplicated, using same des in controlePlane and compute (it has been covered when only setting `ENABLE_DES_DEFAULT_MACHINE`), I will disactive it and remove from 4.15 full function test run.

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-azure

File: ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-ref.yaml [hunk 2]
```diff
@@ -5,6 +5,10 @@ ref:
     name: "4.12"
     tag: upi-installer
   commands: azure-provision-disk-encryption-set-commands.sh
+  env:
+  - name: SET_DIFFERENT_DES_FOR_COMPUTE
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-ref.yaml:9]: we do not  need this ENV any more, right?

File: ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-commands.sh [hunk 3]
```diff
@@ -69,18 +69,58 @@ if [ X"$ret" != X"0" ]; then
...
+    keyvault_key_master="${kv_prefix}-kvkey-master"
+    des_master="${kv_prefix}-des-master"
+    create_disk_encryption_set "${RESOURCE_GROUP}" "${keyvault_master}" "${keyvault_key_master}" "${des_master}"
+
+    echo "Granting service principal reader permissions to the DiskEncryptionSet: ${des_master}"
+    des_master_id=$(az disk-encryption-set show -n "${des_master}" -g "${RESOURCE_GROUP}" --query "[id]" -o tsv)
+    run_command "az role assignment create --assignee ${cluster_sp_id} --role Owner --scope ${des_master_id} -o jsonc"
+
+    #save control plane des information to ${SHARED_DIR} for reference
+    azure_des_json=$("echo ${azure_des_json}" | jq -c -S ". +={\"master\":\"${des_master}\"}")
```
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-commands.sh:104]:
```suggestion
    azure_des_json=$(echo "${azure_des_json}" | jq -c -S ". +={\"master\":\"${des_master}\"}")   
```

File: ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-ref.yaml (hunk 2 above)
Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/disk-encryption-set/azure-provision-disk-encryption-set-ref.yaml]: yes, removed.

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-azure

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-des-minimal-permission-arm-f14

File: ci-operator/step-registry/cucushift/installer/check/azure/des/cucushift-installer-check-azure-des-commands.sh [hunk 4]
```diff
@@ -61,36 +46,73 @@ if [[ -z "${CLUSTER_RESOURCE_GROUP}" ]]; then
...
+
+if [[ "${ENABLE_DES_CONTROL_PLANE}" == "true" ]]; then
+    DES_CONTROL_PLANE=$(cat ${AZURE_DES_FILE} | jq -r '.master')
+fi
+
+if [[ "${ENABLE_DES_COMPUTE}" == "true" ]]; then
+    DES_COMPUTE=$(cat ${AZURE_DES_FILE} | jq -r '.worker')
+fi
+
+ocp_minor_version=$(oc version -o json | jq -r '.openshiftVersion' | cut -d '.' -f2)
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/check/azure/des/cucushift-installer-check-azure-des-commands.sh:65]:
```suggestion
test -f "${SHARED_DIR}/proxy-conf.sh" && source "${SHARED_DIR}/proxy-conf.sh"
ocp_minor_version=$(oc version -o json | jq -r '.openshiftVersion' | cut -d '.' -f2)
```

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-des-minimal-permission-amd-f28-destructive

**Comment by jinyunma** [comment]
test with different des on master and worker is passed in job [periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-des-minimal-permission-amd-f28-destructive](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/47958/rehearse-47958-periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-des-minimal-permission-amd-f28-destructive/1750413947246219264)

As the logical in the step of `ipi-conf-azure-provisioned-des` is changed, it impacts the jobs on 4.11 and 4.12, and set ENV `ENABLE_DES_DEFAULT_MACHINE` explicitly in those jobs.

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.12-multi-nightly-azure-ipi-disk-encryption-set-arm-f28

File: ci-operator/step-registry/ipi/conf/azure/provisioned-des/ipi-conf-azure-provisioned-des-ref.yaml [hunk 5]
```diff
@@ -14,7 +14,16 @@ ref:
...
-      * "true" - Adds provisioned disk encryption set section under DefaultMachinePlatform, which also encrypts the default storage class from 4.13
+  - name: ENABLE_DES_CONTROL_PLANE
+    default: "false"
+    documentation: |-
+      Adds provisioned disk encryption set section under control plane in the install-config.yaml
+  - name: ENABLE_DES_COMPUTE
+    default: "false"
+    documentation: |-
+      Adds provisioned disk encryption set section under control plane in the install-config.yaml
+  - name: SET_DIFFERENT_DES_FOR_COMPUTE
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/provisioned-des/ipi-conf-azure-provisioned-des-ref.yaml:25]: We do not need this env, right?

Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/provisioned-des/ipi-conf-azure-provisioned-des-ref.yaml:25]: We do not need this env, right?

Comment (jinyunma) [reply ci-operator/step-registry/ipi/conf/azure/provisioned-des/ipi-conf-azure-provisioned-des-ref.yaml]: Fixed.

**Comment by jinyunma** [comment]
cc @liangxia to review

**Pull request merged**